packages/react-form/src/hooks/list/dynamiclist.ts

```diff
@@ -0,0 +1,180 @@
+import {useMemo, useEffect, ChangeEvent, useState} from 'react';
+import isEqual from 'fast-deep-equal';
+
+import {
+  ValidationDictionary,
+  Validator,
+  NormalizedValidationDictionary,
+  FieldStates,
+  FieldDictionary,
+  FieldState,
+  ErrorValue,
+  ListValidationContext,
+} from '../../types';
+import {mapObject, normalizeValidation, isChangeEvent} from '../../utilities';
+
+import {
+  updateAction,
+  updateErrorAction,
+  newDefaultAction,
+  reinitializeAction,
+  resetAction,
+  useListReducer,
+  addFieldsAction,
+  removeFieldsAction,
+} from './reducer';
+
+interface DynamicList<Item extends object> {
+  fields: FieldDictionary<Item>[];
+  addField(): void;
+  removeField(index: number): void;
+}
+
+export function useDynamicList<Item extends object>(
+  initialList: Item[],
+  factory: Function,
+  validateFunction?: Partial<
+    ValidationDictionary<Item, ListValidationContext<Item>>
+  >,
+  validationDependencies: unknown[] = [],
+): DynamicList<Item> {
+  const [calculatedList] = useState<Item[]>(initialList);
+
+  const validates = validateFunction ? validateFunction : {};
+  const [state, dispatch] = useListReducer(calculatedList);
+
+  useEffect(() => {
+    if (!isEqual(calculatedList, state.initial)) {
+      dispatch(reinitializeAction(calculatedList));
+    }
+  }, [calculatedList, state.initial, dispatch]);
+
+  const validationConfigs = useMemo(
+    () =>
+      mapObject<NormalizedValidationDictionary<any>>(
+        validates,
+        normalizeValidation,
+      ),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [validates, ...validationDependencies],
+  );
+
+  function addField() {
+    dispatch(addFieldsAction([factory()]));
+  }
+
+  function removeField(index: number) {
+    dispatch(removeFieldsAction(index));
+  }
+
+  const handlers = useMemo(() => {
```

Comment:
I think this would be a good candidate for a custom hook which is shared between the `list` and `dynamic list`. What do you think?

Reply:
I think it's good! Will try to move it to a hook